File: netlify/functions/lib/dataProcessor.mjs
import Papa from 'papaparse';
import * as config from './config.mjs';

function nowStr() {
  return new Date().toISOString().replace('T', ' ').slice(0, 19);
}

export function response(success, data, updatedAt, error) {
  return {
    success,
    updated_at: updatedAt || nowStr(),
    data,
    ...(error ? { error } : {}),
  };
}

function cleanStr(value) {
  if (value == null || value === '') return null;
  const text = String(value).trim();
  return text || null;
}

function emailToName(email) {
  const e = cleanStr(email);
  if (!e || !e.includes('@')) return null;
  const local = e.split('@')[0];
  const parts = local.replace(/\./g, ' ').replace(/_/g, ' ').split(/\s+/);
  return parts.map((p) => p.charAt(0).toUpperCase() + p.slice(1).toLowerCase()).join(' ') || null;
}

function toFloat(value, def = 0) {
  if (value == null || value === '') return def;
  const n = parseFloat(String(value).replace(/,/g, '').trim());
  return Number.isNaN(n) ? def : n;
}

function toInt(value, def = 0) {
  if (value == null || value === '') return def;
  const n = parseInt(String(value).replace(/,/g, '').trim(), 10);
  return Number.isNaN(n) ? def : Math.round(parseFloat(String(value).replace(/,/g, '')) || def);
}

function parseDate(value) {
  if (value == null || value === '') return null;
  const text = String(value).trim();
  if (!text) return null;
  let d;
  if (/^\d{4}/.test(text)) {
    d = new Date(text);
  } else {
    const m = text.match(/^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{2,4})/);
    if (m) {
      let y = parseInt(m[3], 10);
      if (y < 100) y += 2000;
      d = new Date(y, parseInt(m[2], 10) - 1, parseInt(m[1], 10));
    } else {
      d = new Date(text);
    }
  }
  if (Number.isNaN(d.getTime())) return null;
  return d.toISOString().slice(0, 10);
}

function columnPresent(row, candidates) {
  if (!row) return null;
  for (const col of candidates) {
    if (col in row) return col;
  }
  return null;
}

function filterGm(rows) {
  if (!rows.length) return [];
  const gmCol = columnPresent(rows[0], config.GM_COLUMNS);
  if (!gmCol) return [];
  return rows.filter((r) => cleanStr(r[gmCol]) === config.GM_NAME);
}

function paymentMonthColumn(row) {
  // Two "Month" cols in sheet: 1st = label (April-26), 2nd = number (5) — Papa Parse names it Month_1
  if (row && 'Month_1' in row) return 'Month_1';
  if (row && 'Month.1' in row) return 'Month.1';
  if (row && 'Month' in row) return 'Month';
  return null;
}

function rowPaymentMonth(record, monthCol) {
  const raw = record[monthCol];
  if (raw == null || raw === '') return null;
  const n = parseFloat(String(raw).trim().replace(/,/g, ''));
  if (!Number.isNaN(n) && Number.isFinite(n)) return Math.round(n);
  return null;
}

function filterPaymentMonth(rows) {
  if (!rows.length) return [];
  const monthCol = paymentMonthColumn(rows[0]);
  if (!monthCol) return [];
  return rows.filter((r) => rowPaymentMonth(r, monthCol) === config.PAYMENT_MONTH);
}

function normalizeFinalStage(value) {
  const text = cleanStr(value);
  if (!text) return null;
  return text.replace(/ /g, '_');
}

function resolveLeadStage(record) {
  return normalizeFinalStage(record['Final Stage']);
}

async function fetchCsv(name, url) {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), config.REQUEST_TIMEOUT_MS);
  try {
    const res = await fetch(url, { signal: controller.signal });
    if (!res.ok) throw new Error(`CSV ${name} HTTP ${res.status}`);
    const text = await res.text();
    const parsed = Papa.parse(text, { header: true, skipEmptyLines: 'greedy' });
    if (parsed.errors?.length) {
      console.warn(`CSV parse warnings for ${name}:`, parsed.errors.slice(0, 3));
    }
    return (parsed.data || []).filter((row) =>
      Object.values(row).some((v) => v != null && String(v).trim() !== '')
    );
  } finally {
    clearTimeout(t);
  }
}

function buildBdaManagerMap(leadRows) {
  const counts = {};
  for (const r of leadRows) {
    const bda = cleanStr(r['Owner (User Name)']);
    const mgr = cleanStr(r['Manager Name']);
    if (!bda || !mgr) continue;
    if (!counts[bda]) counts[bda] = {};
    counts[bda][mgr] = (counts[bda][mgr] || 0) + 1;
  }
  const map = {};
  for (const [bda, mgrCounts] of Object.entries(counts)) {
    let best = null;
    let bestN = 0;
    for (const [mgr, n] of Object.entries(mgrCounts)) {
      if (n > bestN) {
        bestN = n;
        best = mgr;
      }
    }
    if (best) map[bda] = best;
  }
  return map;
}

function processLeads(rows) {
  const out = [];
  for (const record of rows) {
    const date = parseDate(record['Created On']);
    const mgr = cleanStr(record['Manager Name']);
    const bda = cleanStr(record['Owner (User Name)']);
    const stage = resolveLeadStage(record);
    const source = cleanStr(record.Subsource) || '';
    if (!date || !mgr || !bda || !stage) continue;
    out.push({ date, mgr, bda, stage, source });
  }
  return out;
}

function processInputs(rows, bdaMgr) {
  const out = [];
  for (const record of rows) {
    const date = parseDate(record.Date);
    const bda = emailToName(record['Owner Name']);
    const mgr = cleanStr(record['Manager Name']) || (bda ? bdaMgr[bda] : null);
    if (!date || !bda || !mgr) continue;
    out.push({
      date,
      mgr,
      bda,
      calls: toInt(record['# Calls']),
      connected: toInt(record['# Calls Connected']),
      unique_leads: toInt(record['# Unique Leads']),
      tt: Math.round(toFloat(record['Total Call Duration']) * 100) / 100,
      outbound: toInt(record['# Outbound Calls']),
      inbound: toInt(record['# Inbound Calls']),
      outbound_ans: toInt(record['# Outbound Answered Calls']),
      inbound_ans: toInt(record['# Inbound Answered Calls']),
      dur_lt2: toInt(record['# Call Duration <2mins']),
      dur_2_5: toInt(record['# Call Duration >=2mins & <5mins']),
      dur_5_10: toInt(record['# Call Duration >=5mins & <10mins']),
      dur_10p: toInt(record['# Call Duration >=10mins']),
    });
  }
  return out;
}

function resolveTl(record) {
  for (const col of config.TL_COLUMNS) {
    const tl = cleanStr(record[col]);
    if (tl && config.MANAGERS.includes(tl)) return tl;
  }
  return config.UNASSIGNED_TL;
}

function processTokens(rows) {
  const out = [];
  for (const record of rows) {
    const bda = cleanStr(record.Agent);
    const course = cleanStr(record.Type);
    if (!bda || !course) continue;
    let amount = toFloat(record['Token Amount'], config.DEFAULT_TOKEN_AMOUNT);
    if (cleanStr(record['Token Amount']) == null) amount = config.DEFAULT_TOKEN_AMOUNT;
    out.push({ mgr: resolveTl(record), bda, amount, type: course });
  }
  return out;
}

function processFullPayments(rows) {
  const out = [];
  for (const record of rows) {
    const bda = cleanStr(record.Agent);
    const course = cleanStr(record.Type);
    const amount = toFloat(record['Amount Paid']);
    if (!bda || !course || amount <= 0) continue;
    out.push({ mgr: resolveTl(record), bda, amount, type: course });
  }
  return out;
}

function countByKey(rows, key) {
  const counts = {};
  for (const row of rows) {
    const v = row[key];
    if (!v) continue;
    counts[v] = (counts[v] || 0) + 1;
  }
  return counts;
}

function buildDailyCalls(inputRows) {
  const daily = {};
  for (const row of inputRows) {
    const date = row.date;
    if (!daily[date]) daily[date] = { calls: 0, connected: 0 };
    daily[date].calls += row.calls || 0;
    daily[date].connected += row.connected || 0;
  }
  return daily;
}

function computeKpis(leadRows, inputRows, tokenRows, fullRows) {
  const totalLeads = leadRows.length;
  const totalTokens = tokenRows.length;
  const totalFull = fullRows.length;
  const tokenRevenue = Math.round(tokenRows.reduce((s, r) => s + (r.amount || 0), 0) * 100) / 100;
  const fullRevenue = Math.round(fullRows.reduce((s, r) => s + (r.amount || 0), 0) * 100) / 100;
  return {
    total_leads: totalLeads,
    total_inputs: inputRows.length,
    total_token_payments: totalTokens,
    total_full_payments: totalFull,
    token_revenue: tokenRevenue,
    full_revenue: fullRevenue,
    total_revenue: Math.round((tokenRevenue + fullRevenue) * 100) / 100,
    lead_to_token_conversion_pct: totalLeads ? Math.round((totalTokens / totalLeads) * 10000) / 100 : 0,
    token_to_full_conversion_pct: totalTokens ? Math.round((totalFull / totalTokens) * 10000) / 100 : 0,
    overall_conversion_pct: totalLeads ? Math.round((totalFull / totalLeads) * 10000) / 100 : 0,
    avg_deal_size: totalFull ? Math.round((fullRevenue / totalFull) * 100) / 100 : 0,
  };
}

function buildDashboardPayload(leadRows, inputRows, tokenRows, fullRows) {
  const leadDates = leadRows.map((r) => r.date);
  const inputDates = inputRows.map((r) => r.date);
  const allDates = [...leadDates, ...inputDates];
  const dateMin = allDates.length ? allDates.reduce((a, b) => (a < b ? a : b)) : config.MONTH_START;
  const dateMax = allDates.length ? allDates.reduce((a, b) => (a > b ? a : b)) : config.MONTH_END;
  return {
    gm: config.GM_NAME,
    lead_rows: leadRows,
    input_rows: inputRows,
    token_rows: tokenRows,
    full_rows: fullRows,
    full_by_course: countByKey(fullRows, 'type'),
    token_by_course: countByKey(tokenRows, 'type'),
    leads_by_source: countByKey(leadRows, 'source'),
    daily_calls: buildDailyCalls(inputRows),
    date_min: dateMin,
    date_max: dateMax,
    input_date_min: inputDates.length ? inputDates.reduce((a, b) => (a < b ? a : b)) : null,
    input_date_max: inputDates.length ? inputDates.reduce((a, b) => (a > b ? a : b)) : null,
  };
}

let cache = null;
let cacheAt = null;

export async function refreshStore() {
  config.validateConfig();
  const frames = {};
  for (const [name, url] of Object.entries(config.CSV_URLS)) {
    frames[name] = await fetchCsv(name, url);
  }

  const leadsDf = filterGm(frames.leads);
  const inputsDf = filterGm(frames.inputs);
  const tokensDf = filterPaymentMonth(filterGm(frames.tokens));
  const fullDf = filterPaymentMonth(filterGm(frames.fullpayments));

  const bdaMgr = buildBdaManagerMap(leadsDf);
  const leadRows = processLeads(leadsDf);
  const inputRows = processInputs(inputsDf, bdaMgr);
  const tokenRows = processTokens(tokensDf);
  const fullRows = processFullPayments(fullDf);

  const dashboard = buildDashboardPayload(leadRows, inputRows, tokenRows, fullRows);
  const kpis = computeKpis(leadRows, inputRows, tokenRows, fullRows);

  cacheAt = nowStr();
  cache = { dashboard, kpis, leadRows, inputRows, tokenRows, fullRows };
  return cache;
}

export function getCache() {
  return { cache, cacheAt };
}

File: data_processor.py
"""Fetch, clean, filter, and aggregate Google Sheets CSV data for the dashboard."""

from __future__ import annotations

import io
import logging
from datetime import datetime
from typing import Any

import pandas as pd
import requests

import config

logger = logging.getLogger(__name__)


def _response(success: bool, data: Any, updated_at: str | None = None, error: str | None = None) -> dict:
    payload: dict[str, Any] = {
        "success": success,
        "updated_at": updated_at or datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "data": data,
    }
    if error:
        payload["error"] = error
    return payload


def _clean_str(value: Any) -> str | None:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return None
    text = str(value).strip()
    return text or None


def _email_to_name(email: Any) -> str | None:
    email = _clean_str(email)
    if not email or "@" not in email:
        return None
    local = email.split("@", 1)[0]
    parts = local.replace(".", " ").replace("_", " ").split()
    return " ".join(part.capitalize() for part in parts) if parts else None


def _to_float(value: Any, default: float = 0.0) -> float:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return default
    try:
        return float(str(value).replace(",", "").strip())
    except (TypeError, ValueError):
        return default


def _to_int(value: Any, default: int = 0) -> int:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return default
    try:
        return int(float(value))
    except (TypeError, ValueError):
        return default


def _parse_date(value: Any) -> str | None:
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return None
    text = str(value).strip()
    if not text:
        return None
    dayfirst = not text[:4].isdigit()
    parsed = pd.to_datetime(value, errors="coerce", dayfirst=dayfirst)
    if pd.isna(parsed):
        return None
    return parsed.strftime("%Y-%m-%d")


def _drop_blank_rows(df: pd.DataFrame) -> pd.DataFrame:
    if df.empty:
        return df
    return df.dropna(how="all").copy()


def _column_present(df: pd.DataFrame, candidates: tuple[str, ...]) -> str | None:
    for col in candidates:
        if col in df.columns:
            return col
    return None


def _filter_gm(df: pd.DataFrame) -> pd.DataFrame:
    """Keep rows where GM column equals Siddhartha (all managers' payments under this GM)."""
    gm_col = _column_present(df, config.GM_COLUMNS)
    if not gm_col:
        return df.iloc[0:0].copy()
    gm = df[gm_col].astype(str).str.strip()
    return df[gm == config.GM_NAME].copy()


def _filter_lead_dates(df: pd.DataFrame) -> pd.DataFrame:
    """Keep leads within the dashboard month window (May 1–20)."""
    if df.empty or "Created On" not in df.columns:
        return df.iloc[0:0].copy()
    dates = pd.to_datetime(df["Created On"], errors="coerce")
    mask = (dates >= config.LEAD_DATE_START) & (dates <= config.LEAD_DATE_END)
    return df[mask].copy()


def _normalize_final_stage(value: Any) -> str | None:
    text = _clean_str(value)
    if not text:
        return None
    # Final Stage uses underscores (e.g. Not_Connected, Interested-Test)
    return text.replace(" ", "_")


# CRM Stage column maps to dashboard buckets (matches original embedded dashboard logic)
_STAGE_TO_BUCKET: dict[str, str] = {
    "Warm": "Warm",
    "Follow Up": "Follow_Up",
    "Call Back Later": "Call_Back_Later",
    "Not Interested": "Not_Interested",
    "Not Eligible": "Not_Eligible",
    "Invalid": "Invalid",
    "Interested": "Interested",
    "Token Paid": "Token_Paid",
    "Did Not Picked": "Not_Connected",
    "Never Picked Up": "Not_Connected",
    "Did Not Enquired": "Not_Connected",
    "Call Not Connected": "Not_Connected",
    "Wrong Number": "Invalid",
    "Future Prospect": "Follow_Up",
    "Test Sent": "Follow_Up",
    "New Enquiry": "New_Enquiry",
    "Learner Enrolled": "Learner_Enrolled",
}


def resolve_lead_stage(record: Any) -> str | None:
    """Lead funnel and stage KPIs use Final Stage column only."""
    return _normalize_final_stage(record.get("Final Stage"))


def _filter_payment_month(df: pd.DataFrame) -> pd.DataFrame:
    if df.empty:
        return df
    month_col = "Month.1" if "Month.1" in df.columns else "Month"
    if month_col not in df.columns:
        return df.iloc[0:0].copy()
    month_values = pd.to_numeric(df[month_col], errors="coerce")
    return df[month_values == config.PAYMENT_MONTH].copy()


def fetch_csv(name: str, url: str) -> pd.DataFrame:
    """Download a published Google Sheet CSV into a cleaned DataFrame."""
    logger.info("Fetching CSV: %s", name)
    response = requests.get(url, timeout=config.REQUEST_TIMEOUT)
    response.raise_for_status()
    df = pd.read_csv(io.StringIO(response.text), low_memory=False)
    df = _drop_blank_rows(df)
    logger.info("Loaded %s rows for %s", len(df), name)
    return df


def build_bda_manager_map(lead_df: pd.DataFrame) -> dict[str, str]:
    """Map BDA names to manager names using lead ownership data."""
    if lead_df.empty:
        return {}

    subset = lead_df.dropna(subset=["Owner (User Name)", "Manager Name"]).copy()
    if subset.empty:
        return {}

    grouped = (
        subset.groupby("Owner (User Name)")["Manager Name"]
        .agg(lambda values: values.mode().iloc[0] if not values.mode().empty else values.iloc[0])
    )
    return {str(k).strip(): str(v).strip() for k, v in grouped.items() if _clean_str(k) and _clean_str(v)}


def process_leads(df: pd.DataFrame) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for _, record in df.iterrows():
        date = _parse_date(record.get("Created On"))
        mgr = _clean_str(record.get("Manager Name"))
        bda = _clean_str(record.get("Owner (User Name)"))
        stage = resolve_lead_stage(record)
        source = _clean_str(record.get("Subsource")) or ""

        if not date or not mgr or not bda or not stage:
            continue

        rows.append(
            {
                "date": date,
                "mgr": mgr,
                "bda": bda,
                "stage": stage,
                "source": source,
            }
        )
    return rows


def process_inputs(df: pd.DataFrame, bda_mgr: dict[str, str]) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for _, record in df.iterrows():
        date = _parse_date(record.get("Date"))
        bda = _email_to_name(record.get("Owner Name"))
        mgr = _clean_str(record.get("Manager Name")) or (bda_mgr.get(bda) if bda else None)

        if not date or not bda or not mgr:
            continue

        rows.append(
            {
                "date": date,
                "mgr": mgr,
                "bda": bda,
                "calls": _to_int(record.get("# Calls")),
                "connected": _to_int(record.get("# Calls Connected")),
                "unique_leads": _to_int(record.get("# Unique Leads")),
                "tt": round(_to_float(record.get("Total Call Duration")), 2),
                "outbound": _to_int(record.get("# Outbound Calls")),
                "inbound": _to_int(record.get("# Inbound Calls")),
                "outbound_ans": _to_int(record.get("# Outbound Answered Calls")),
                "inbound_ans": _to_int(record.get("# Inbound Answered Calls")),
                "dur_lt2": _to_int(record.get("# Call Duration <2mins")),
                "dur_2_5": _to_int(record.get("# Call Duration >=2mins & <5mins")),
                "dur_5_10": _to_int(record.get("# Call Duration >=5mins & <10mins")),
                "dur_10p": _to_int(record.get("# Call Duration >=10mins")),
            }
        )
    return rows


def _resolve_tl(record: Any) -> str:
    """
    Team lead (manager) from TL column only.

    Do not infer from lead dump / Agent — future rows may have GM=Siddhartha
    with an empty TL; those still count in KPI totals as Unassigned.
    """
    for col in config.TL_COLUMNS:
        tl = _clean_str(record.get(col))
        if tl in config.MANAGERS:
            return tl
    return config.UNASSIGNED_TL


def process_tokens(df: pd.DataFrame) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for _, record in df.iterrows():
        bda = _clean_str(record.get("Agent"))
        course = _clean_str(record.get("Type"))
        if not bda or not course:
            continue

        mgr = _resolve_tl(record)
        amount = _to_float(record.get("Token Amount"), default=config.DEFAULT_TOKEN_AMOUNT)
        if pd.isna(record.get("Token Amount")) or _clean_str(record.get("Token Amount")) is None:
            amount = config.DEFAULT_TOKEN_AMOUNT

        rows.append(
            {
                "mgr": mgr,
                "bda": bda,
                "amount": amount,
                "type": course,
            }
        )
    return rows


def process_full_payments(df: pd.DataFrame) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for _, record in df.iterrows():
        bda = _clean_str(record.get("Agent"))
        course = _clean_str(record.get("Type"))
        amount = _to_float(record.get("Amount Paid"))
        if not bda or not course or amount <= 0:
            continue

        mgr = _resolve_tl(record)
        rows.append(
            {
                "mgr": mgr,
                "bda": bda,
                "amount": amount,
                "type": course,
            }
        )
    return rows


def count_by_key(rows: list[dict[str, Any]], key: str) -> dict[str, int]:
    counts: dict[str, int] = {}
    for row in rows:
        value = row.get(key)
        if not value:
            continue
        counts[value] = counts.get(value, 0) + 1
    return counts


def aggregate_stages(lead_rows: list[dict[str, Any]]) -> dict[str, int]:
    return count_by_key(lead_rows, "stage")


def build_daily_calls(input_rows: list[dict[str, Any]]) -> dict[str, dict[str, int]]:
    daily: dict[str, dict[str, int]] = {}
    for row in input_rows:
        date = row["date"]
        if date not in daily:
            daily[date] = {"calls": 0, "connected": 0}
        daily[date]["calls"] += row.get("calls", 0)
        daily[date]["connected"] += row.get("connected", 0)
    return daily


def compute_kpis(
    lead_rows: list[dict[str, Any]],
    input_rows: list[dict[str, Any]],
    token_rows: list[dict[str, Any]],
    full_rows: list[dict[str, Any]],
) -> dict[str, Any]:
    total_leads = len(lead_rows)
    total_inputs = len(input_rows)
    total_tokens = len(token_rows)
    total_full = len(full_rows)
    token_revenue = round(sum(row.get("amount", 0) for row in token_rows), 2)
    full_revenue = round(sum(row.get("amount", 0) for row in full_rows), 2)
    total_revenue = round(token_revenue + full_revenue, 2)

    lead_to_token = round((total_tokens / total_leads * 100), 2) if total_leads else 0.0
    token_to_full = round((total_full / total_tokens * 100), 2) if total_tokens else 0.0
    overall = round((total_full / total_leads * 100), 2) if total_leads else 0.0

    return {
        "total_leads": total_leads,
        "total_inputs": total_inputs,
        "total_token_payments": total_tokens,
        "total_full_payments": total_full,
        "token_revenue": token_revenue,
        "full_revenue": full_revenue,
        "total_revenue": total_revenue,
        "lead_to_token_conversion_pct": lead_to_token,
        "token_to_full_conversion_pct": token_to_full,
        "overall_conversion_pct": overall,
        "avg_deal_size": round(full_revenue / total_full, 2) if total_full else 0.0,
    }


def build_funnel(lead_rows: list[dict[str, Any]]) -> dict[str, Any]:
    stages = aggregate_stages(lead_rows)
    total = len(lead_rows)
    ordered = [
        "Interested",
        "Follow_Up",
        "Call_Back_Later",
        "Token_Paid",
        "Learner_Enrolled",
        "Fresh_Lead",
        "New_Enquiry",
        "Not_Connected",
        "Not_Interested",
        "Not_Eligible",
        "Invalid",
    ]
    funnel = []
    for stage in ordered:
        count = stages.get(stage, 0)
        funnel.append(
            {
                "stage": stage,
                "count": count,
                "pct": round((count / total * 100), 2) if total else 0.0,
            }
        )
    return {"total_leads": total, "stages": funnel, "stage_counts": stages}


def build_revenue(token_rows: list[dict[str, Any]], full_rows: list[dict[str, Any]]) -> dict[str, Any]:
    token_revenue = round(sum(row.get("amount", 0) for row in token_rows), 2)
    full_revenue = round(sum(row.get("amount", 0) for row in full_rows), 2)
    return {
        "token_count": len(token_rows),
        "full_count": len(full_rows),
        "token_revenue": token_revenue,
        "full_revenue": full_revenue,
        "total_revenue": round(token_revenue + full_revenue, 2),
        "full_by_course": count_by_key(full_rows, "type"),
        "token_by_course": count_by_key(token_rows, "type"),
        "by_manager": {
            mgr: {
                "full_count": sum(1 for row in full_rows if row.get("mgr") == mgr),
                "full_revenue": round(sum(row.get("amount", 0) for row in full_rows if row.get("mgr") == mgr), 2),
                "token_count": sum(1 for row in token_rows if row.get("mgr") == mgr),
                "token_revenue": round(sum(row.get("amount", 0) for row in token_rows if row.get("mgr") == mgr), 2),
            }
            for mgr in config.MANAGERS
        },
    }


def build_charts(
    lead_rows: list[dict[str, Any]],
    input_rows: list[dict[str, Any]],
    token_rows: list[dict[str, Any]],
    full_rows: list[dict[str, Any]],
) -> dict[str, Any]:
    return {
        "leads_by_source": count_by_key(lead_rows, "source"),
        "full_by_course": count_by_key(full_rows, "type"),
        "token_by_course": count_by_key(token_rows, "type"),
        "daily_calls": build_daily_calls(input_rows),
        "daily_leads": count_by_key(lead_rows, "date"),
        "stage_by_manager": {
            mgr: aggregate_stages([row for row in lead_rows if row.get("mgr") == mgr])
            for mgr in config.MANAGERS
        },
    }


def build_tables(
    lead_rows: list[dict[str, Any]],
    input_rows: list[dict[str, Any]],
    token_rows: list[dict[str, Any]],
    full_rows: list[dict[str, Any]],
) -> dict[str, Any]:
    bda_revenue: dict[str, dict[str, Any]] = {}

    for row in full_rows:
        name = row["bda"]
        if name not in bda_revenue:
            bda_revenue[name] = {
                "bda": name,
                "mgr": row.get("mgr", ""),
                "full_count": 0,
                "full_amount": 0.0,
                "token_count": 0,
                "token_amount": 0.0,
            }
        bda_revenue[name]["full_count"] += 1
        bda_revenue[name]["full_amount"] += row.get("amount", 0)

    for row in token_rows:
        name = row["bda"]
        if name not in bda_revenue:
            bda_revenue[name] = {
                "bda": name,
                "mgr": row.get("mgr", ""),
                "full_count": 0,
                "full_amount": 0.0,
                "token_count": 0,
                "token_amount": 0.0,
            }
        bda_revenue[name]["token_count"] += 1
        bda_revenue[name]["token_amount"] += row.get("amount", 0)

    bda_productivity: dict[str, dict[str, Any]] = {}
    for row in input_rows:
        name = row["bda"]
        if name not in bda_productivity:
            bda_productivity[name] = {
                "bda": name,
                "mgr": row.get("mgr", ""),
                "calls": 0,
                "connected": 0,
                "unique_leads": 0,
                "tt": 0.0,
                "days": set(),
            }
        entry = bda_productivity[name]
        entry["calls"] += row.get("calls", 0)
        entry["connected"] += row.get("connected", 0)
        entry["unique_leads"] += row.get("unique_leads", 0)
        entry["tt"] += row.get("tt", 0)
        entry["days"].add(row.get("date"))

    productivity_rows = []
    for entry in bda_productivity.values():
        productivity_rows.append(
            {
                "bda": entry["bda"],
                "mgr": entry["mgr"],
                "calls": entry["calls"],
                "connected": entry["connected"],
                "unique_leads": entry["unique_leads"],
                "tt": round(entry["tt"], 2),
                "days": len(entry["days"]),
            }
        )

    bda_leads: dict[str, dict[str, Any]] = {}
    for row in lead_rows:
        key = f"{row['mgr']}|{row['bda']}"
        if key not in bda_leads:
            bda_leads[key] = {"mgr": row["mgr"], "bda": row["bda"], "stages": {}}
        stage = row.get("stage", "")
        bda_leads[key]["stages"][stage] = bda_leads[key]["stages"].get(stage, 0) + 1

    return {
        "bda_revenue": sorted(bda_revenue.values(), key=lambda item: item["full_amount"], reverse=True),
        "bda_productivity": sorted(productivity_rows, key=lambda item: item["calls"], reverse=True),
        "bda_leads": list(bda_leads.values()),
    }


def build_dashboard_payload(
    lead_rows: list[dict[str, Any]],
    input_rows: list[dict[str, Any]],
    token_rows: list[dict[str, Any]],
    full_rows: list[dict[str, Any]],
) -> dict[str, Any]:
    """Build the RAW-compatible object consumed by the existing frontend."""
    lead_dates = [row["date"] for row in lead_rows]
    input_dates = [row["date"] for row in input_rows]
    all_dates = lead_dates + input_dates
    date_min = min(all_dates) if all_dates else config.MONTH_START
    date_max = max(all_dates) if all_dates else config.MONTH_END

    return {
        "gm": config.GM_NAME,
        "lead_rows": lead_rows,
        "input_rows": input_rows,
        "token_rows": token_rows,
        "full_rows": full_rows,
        "full_by_course": count_by_key(full_rows, "type"),
        "token_by_course": count_by_key(token_rows, "type"),
        "leads_by_source": count_by_key(lead_rows, "source"),
        "daily_calls": build_daily_calls(input_rows),
        "date_min": date_min,
        "date_max": date_max,
        "input_date_min": min(input_dates) if input_dates else None,
        "input_date_max": max(input_dates) if input_dates else None,
    }


class DataStore:
    """In-memory cache for processed dashboard datasets."""

    def __init__(self) -> None:
        self.updated_at: str | None = None
        self.last_error: str | None = None
        self.raw_frames: dict[str, pd.DataFrame] = {}
        self.lead_rows: list[dict[str, Any]] = []
        self.input_rows: list[dict[str, Any]] = []
        self.token_rows: list[dict[str, Any]] = []
        self.full_rows: list[dict[str, Any]] = []
        self.dashboard: dict[str, Any] = {}
        self.kpis: dict[str, Any] = {}
        self.funnel: dict[str, Any] = {}
        self.revenue: dict[str, Any] = {}
        self.charts: dict[str, Any] = {}
        self.tables: dict[str, Any] = {}

    def refresh(self) -> None:
        """Fetch all CSVs, clean, filter, and rebuild cached aggregates."""
        frames: dict[str, pd.DataFrame] = {}
        for name, url in config.CSV_URLS.items():
            frames[name] = fetch_csv(name, url)

        # All Siddhartha leads from sheet; date filtering is applied in the frontend
        leads_df = _filter_gm(frames["leads"])
        inputs_df = _filter_gm(frames["inputs"])
        tokens_df = _filter_payment_month(_filter_gm(frames["tokens"]))
        full_df = _filter_payment_month(_filter_gm(frames["fullpayments"]))

        bda_mgr = build_bda_manager_map(leads_df)

        self.lead_rows = process_leads(leads_df)
        self.input_rows = process_inputs(inputs_df, bda_mgr)
        self.token_rows = process_tokens(tokens_df)
        self.full_rows = process_full_payments(full_df)

        self.dashboard = build_dashboard_payload(
            self.lead_rows,
            self.input_rows,
            self.token_rows,
            self.full_rows,
        )
        self.kpis = compute_kpis(self.lead_rows, self.input_rows, self.token_rows, self.full_rows)
        self.funnel = build_funnel(self.lead_rows)
        self.revenue = build_revenue(self.token_rows, self.full_rows)
        self.charts = build_charts(self.lead_rows, self.input_rows, self.token_rows, self.full_rows)
        self.tables = build_tables(self.lead_rows, self.input_rows, self.token_rows, self.full_rows)

        self.raw_frames = {
            "leads": leads_df,
            "inputs": inputs_df,
            "tokens": tokens_df,
            "fullpayments": full_df,
        }
        self.updated_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        self.last_error = None
        logger.info(
            "Cache refreshed: leads=%s inputs=%s tokens=%s full=%s",
            len(self.lead_rows),
            len(self.input_rows),
            len(self.token_rows),
            len(self.full_rows),
        )

    def as_response(self, data: Any) -> dict[str, Any]:
        return _response(True, data, updated_at=self.updated_at)

    def error_response(self, message: str) -> dict[str, Any]:
        return _response(False, None, updated_at=self.updated_at, error=message)
